server/src/main/java/org/elasticsearch/cluster/coordination/MasterHistory.java

```diff
@@ -0,0 +1,215 @@
+     * node1 -> null -> node2 is 1 identity change
+     * @param masterHistory The list of nodes that have been master
+     * @return The number of master identity changes as defined above
+     */
+    public static int getNumberOfMasterIdentityChanges(List<DiscoveryNode> masterHistory) {
+        int identityChanges = 0;
+        List<DiscoveryNode> nonNullHistory = masterHistory.stream().filter(Objects::nonNull).toList();
+        DiscoveryNode previousNode = null;
+        for (DiscoveryNode node : nonNullHistory) {
+            if (previousNode != null && previousNode.equals(node) == false) {
+                identityChanges++;
+            }
+            previousNode = node;
+        }
+        return identityChanges;
+    }
+
+    /**
+     * Returns true if a non-null master was seen at any point in the last n seconds, or if the last-seen master was more than n seconds
+     * ago and non-null.
+     * @param n The number of seconds to look back
+     * @return true if the current master is non-null or if a non-null master was seen in the last n seconds
+     */
+    public boolean hasSeenMasterInLastNSeconds(int n) {
+        List<TimeAndMaster> masterHistoryCopy = getRecentMasterHistory(masterHistory);
+        long now = nowSupplier.get();
+        TimeValue nSeconds = new TimeValue(n, TimeUnit.SECONDS);
+        long nSecondsAgo = now - nSeconds.getMillis();
+        return getMostRecentMaster() != null
+            || masterHistoryCopy.stream()
+                .filter(timeAndMaster -> timeAndMaster.master != null)
+                .anyMatch(timeAndMaster -> timeAndMaster.time > nSecondsAgo);
+    }
+
+    /*
+     * This method creates a copy of masterHistory that only has entries from more than MAX_HISTORY_AGE before now (but leaves the newest
+     * entry in even if it is more than MAX_HISTORY_AGE).
+     */
+    private List<TimeAndMaster> getRecentMasterHistory(List<TimeAndMaster> history) {
+        if (history.size() < 2) {
+            return history;
+        }
+        long now = nowSupplier.get();
+        long oldestRelevantHistoryTime = now - MAX_HISTORY_AGE.getMillis();
+        TimeAndMaster mostRecent = history.isEmpty() ? null : history.get(history.size() - 1);
+        List<TimeAndMaster> filteredHistory = history.stream()
+            .filter(timeAndMaster -> timeAndMaster.time > oldestRelevantHistoryTime)
+            .collect(Collectors.toList());
+        if (filteredHistory.isEmpty() && mostRecent != null) { // The most recent entry was more than 30 minutes ago
+            filteredHistory.add(mostRecent);
+        }
+        return filteredHistory;
+    }
+
+    /**
+     * Returns a new MasterHistory with nothing from more than MAX_HISTORY_AGE before now (but leaves the newest entry in even if it is
+     * more than MAX_HISTORY_AGE).
+     */
+    private List<TimeAndMaster> removeOldMasterHistory(List<TimeAndMaster> possiblyOldMasterHistory) {
```

Comment:
I wonder can we prune more things here? We should be able to determine that the master is unstable with just a few items, we don't need to keep thousands of them around. Maybe we can set an upper bound on the list length?